src/builtin.c

```diff
@@ -807,6 +806,63 @@ static jv f_sort_by_impl(jq_state *jq, jv input, jv keys) {
   }
 }
 
+/* Assuming the input array is sorted, bsearch/1 returns */
+/* the index of the target if the target is in the input array; and otherwise */
+/*  (-1 - ix), where ix is the insertion point that would leave the array sorted. */
+/* If the input is not sorted, bsearch will terminate but with irrelevant results. */
+static jv f_bsearch(jq_state *jq, jv input, jv target) {
+    assert(jv_get_kind(input) == JV_KIND_ARRAY);
+    assert(jv_get_kind(target) == JV_KIND_NUMBER);
+    int len = jv_array_length(jv_copy(input));
+    if (len == 0) {
+        jv_free(input);
+        jv_free(target);
+        return jv_number(-1);
+    } else if (len == 1) {
+        int result = jv_cmp(target, jv_array_get(jv_copy(input), 0));
+        jv_free(input);
+        if (result == 0 ) {
+            return jv_number(0);
+        } else if (result > 0) {
+            return jv_number(-2);
+        } else {
+            return jv_number(-1);
+        }
+    }
+
+    int start = 0;
+    int end = len - 1;
+    jv answer = jv_null();
+    while (start <end) {
+        int mid = (start + end) / 2;
+        int result = jv_cmp(jv_copy(target), jv_array_get(jv_copy(input), mid));
+        if (result == 0) {
+            answer = jv_number(mid);
+            break;
+        } else if (start == end ) {
+            answer = jv_number(-1);
+            break;
+        } else if (result < 0 ) {
+            end = mid -1;
+        } else {
+            start = mid +1;
+        }
+    }
+
+    if (jv_equal(answer, jv_null())) {
```

Comment:
this `jv_equal` will free `answer` which is not good, since you want to return it if its not `null`, you should have copied it.

Also the best way to check if something is `null` is with `if (jv_get_kind(answer) == JV_KIND_NULL)`, this is one of the few `jv_` functions that does not free its argument automatically so no need to copy if you use that.

Reply:
I guess it is working out anyway because `answer` is always either `jv_null()` or a `jv_number()` which are never allocated.